src/main/java/io/nats/client/support/JwtUtils.java

```diff
@@ -0,0 +1,145 @@
+            "------END USER NKEY SEED------\n" +
+            "\n" +
+            "*************************************************************";
+
+    public static String issueUserCreds(String accountSeed, String issuerAccount, String userSeed, String userKeyPub, String optionalName, Duration optionalExpiration, String... optionalTags) throws GeneralSecurityException, IOException {
+        String jwt = issueUserJWT(accountSeed, issuerAccount, userKeyPub, optionalName, optionalExpiration, optionalTags);
+        return String.format(NATS_USER_JWT_FORMAT, jwt, userSeed);
+    }
+
+    public static String issueUserCreds(NKey accountSigningKey, String issuerAccount, String userSeed, String userKeyPub, String optionalName, Duration optionalExpiration, String... optionalTags) throws GeneralSecurityException, IOException {
+        String jwt = issueUserJWT(accountSigningKey, issuerAccount, userKeyPub, optionalName, optionalExpiration, optionalTags);
+        return String.format(NATS_USER_JWT_FORMAT, jwt, userSeed);
+    }
+
+    public static String issueUserJWT(String accountSeed, String issuerAccount, String userKeyPub, String optionalName, Duration optionalExpiration, String... optionalTags) throws GeneralSecurityException, IOException {
+        NKey accountSigningKey = NKey.fromSeed(accountSeed.toCharArray());
+        return issueUserJWT(accountSigningKey, issuerAccount, userKeyPub, optionalName, optionalExpiration, optionalTags);
+    }
+
+    public static String issueUserJWT(NKey accountSigningKey, String issuerAccount, String userKeyPub, String optionalName, Duration optionalExpiration, String... optionalTags) throws GeneralSecurityException, IOException {
+        String accSigningKeyPub = new String(accountSigningKey.getPublicKey());
+
+        Claim claim = new Claim();
+        claim.exp = optionalExpiration;
+        claim.iat = System.currentTimeMillis() / 1000;
+        claim.iss = accSigningKeyPub;
+        claim.name = Validator.nullOrEmpty(optionalName) ? userKeyPub : optionalName;
+        claim.sub = userKeyPub;
+        claim.nats = new Nats();
+        claim.nats.issuerAccount = issuerAccount;
+        claim.nats.tags = optionalTags;
+
+        // Issue At time is stored in unix seconds
+        String claimJson = claim.toJson();
+
+        // Compute jti, a base32 encoded sha256 hash
+        MessageDigest sha256 = MessageDigest.getInstance("SHA-256");
+        byte[] encoded = sha256.digest(claimJson.getBytes(StandardCharsets.UTF_8));
+
+        claim.jti = new String(NKey.base32Encode(encoded));
+        claimJson = claim.toJson();
+
+        // all three components (header/body/signature) are base64url encoded
+        String encBody = ToBase64Url(claimJson);
+
+        // compute the signature off of header + body (. included on purpose)
+        byte[] sig = (ENCODED_CLAIM_HEADER + "." + encBody).getBytes(StandardCharsets.UTF_8);
+        String encSig = ToBase64Url(accountSigningKey.sign(sig));
```

Comment:
I have similar JWT issuing code, but the signature is computed off the `encBody` only and does NOT include the ENCODED_CLAIM_HEADER or dot. Did something change about the spec recently?

More specifically, I thought this is what the code needed to be:

```java
String encSig = ToBase64Url(accountSigningKey.sign(encBody));
```

...but I also use the `ed25519` "alg" in the JWT header, and NOT `ed25519-nkey`. So, perhaps the spec changed between these two algorithms?

Reply:
I see that https://github.com/nats-io/nats-architecture-and-design/blob/main/adr/ADR-14.md explicitly mentions this new way of computing the signature, so I'm assuming this is just the design of the new `ed25519-nkey` algorithm.

Reply:
defer to Matthias